src/box/memtx_space.c

```diff
@@ -1125,7 +1125,11 @@ memtx_init_ephemeral_space(struct space *space)
  * corresponding on_rollback triggers to prevent rollbacked changes appearance.
  */
 struct index_build_on_rollback_data {
-	struct memtx_ddl_state *state;
+	/*
+	 * Weak reference to the index. Cannot store usual reference
+	 * since the index can be altered before rollback.
+	 */
+	struct index_weak_ref index_ref;
```

Comment:
Using `index_weak_ref` for this is an overkill IMO. Let's just store `index->def->iid` and look up the index in `txn_stmt::space`.

Reply:
I realised that it doesn't work - space cache contains the old space (the new one will be put there only when the new index is built). It worked only because `space_cache_version` wasn't updated and the `index_weak_ref` used to passed pointer and didn't lookup in the space cache.

I completely reworked the patch - the new approach is used.